tachyon/math/finite_fields/binary_field.h

```diff
@@ -0,0 +1,299 @@
+#include <type_traits>
+
+#include "tachyon/base/logging.h"
+#include "tachyon/base/numerics/safe_conversions.h"
+#include "tachyon/base/random.h"
+#include "tachyon/base/strings/string_number_conversions.h"
+#include "tachyon/base/strings/string_util.h"
+#include "tachyon/build/build_config.h"
+#include "tachyon/math/finite_fields/binary_field_traits_forward.h"
+#include "tachyon/math/finite_fields/binary_tower_operations.h"
+#include "tachyon/math/finite_fields/finite_field.h"
+
+namespace tachyon::math {
+
+template <typename _Config>
+class BinaryField final : public FiniteField<BinaryField<_Config>> {
+ public:
+  constexpr static size_t kBits = _Config::kModulusBits - 1;
+  constexpr static size_t kLimbNums = (kBits + 63) / 64;
+  constexpr static size_t N = kLimbNums;
+
+  using Config = _Config;
+  using Type = std::conditional_t<
+      kBits <= 8, uint8_t,
+      std::conditional_t<
+          kBits == 16, uint16_t,
+          std::conditional_t<
+              kBits == 32, uint32_t,
+              std::conditional_t<kBits == 64, uint64_t, BigInt<2>>>>>;
+  using SubField = BinaryField<typename BinaryFieldTraits<Config>::SubConfig>;
+  using BigIntTy = BigInt<N>;
+  using value_type = Type;
+
+  using CpuField = BinaryField<Config>;
+  using GpuField = BinaryField<Config>;
+
+  constexpr BinaryField() = default;
+  template <typename T, std::enable_if_t<std::is_integral_v<T>>* = nullptr>
+  constexpr explicit BinaryField(T value) {
+    DCHECK_GE(value, T{0});
+    if constexpr (kBits <= 64) {
+      DCHECK(base::IsValueInRangeForNumericType<Type>(value));
+      DCHECK_LE(static_cast<Type>(value), GetMax());
+      value_ = value;
+    } else {
+      value_[0] = value;
+    }
+  }
+  constexpr explicit BinaryField(BigInt<1> value) : BinaryField(value[0]) {
+    static_assert(kBits <= 64);
+  }
+  constexpr explicit BinaryField(BigInt<2> value) : value_(value) {
+    static_assert(kBits == 128);
+  }
+  constexpr BinaryField(const BinaryField& other) = default;
+  constexpr BinaryField& operator=(const BinaryField& other) = default;
+  constexpr BinaryField(BinaryField&& other) = default;
+  constexpr BinaryField& operator=(BinaryField&& other) = default;
+
+  constexpr static BinaryField Zero() { return BinaryField(); }
+  constexpr static BinaryField One() { return BinaryField(1); }
+
+  static BinaryField Random() {
+    if constexpr (kBits <= 64) {
+      return BinaryField(
+          base::Uniform(base::Range<Type, true, true>::Until(GetMax())));
+    } else {
+      static_assert(kBits == 128);
+      return BinaryField(BigInt<2>::Random());
+    }
+  }
+
+  static std::optional<BinaryField> FromDecString(std::string_view str) {
+    if constexpr (kBits <= 64) {
+      using MaybePromotedType =
+          std::conditional_t<(kBits <= 32), uint32_t, Type>;
+      MaybePromotedType value;
+      if (!absl::SimpleAtoi(str, &value)) return std::nullopt;
+      if constexpr (kBits <= 16) {
+        if (value > GetMax()) {
```

Comment:
Why is value not checked for `kBits > 16`?

Reply:
See https://github.com/kroma-network/tachyon/pull/416#discussion_r1623798467